Migrate to typescript

stefantds opened the pull request:
This PR is migrating the library to typescript **almost** without any other changes (the changes introduced are highlighted by comments).
As a follow-up we should continue making the library more typescript-y (e.g. replace multiple functions `decrypt[Object|String|etc]` with union types. This was not in scope for this refactoring.

stefantds commented on .eslintrc.js:
```diff
@@ -1,35 +1,19 @@
         'import',
+        '@typescript-eslint'
     ],
-    rules: {
```
I used this occasion to simplify the linting rules. If there are opipnions about custom rules I'm happy to add them back but the recommended values are quite a good start.

stefantds commented on .eslintrc.js:
```diff
@@ -1,35 +1,19 @@
 module.exports = {
-    extends: ['eslint:recommended', 'airbnb-base'],
```
We could use the airbnb typescript rules set but I am slightly in favour of the more official rules.

radibit replied:
Yes, completely agree 👍

stefantds commented on karma.conf.js:
```diff
@@ -6,19 +6,13 @@ module.exports = function (config) {
-            'source/**/*.js': ['browserify'],
-            'test/**/*.js': ['browserify'],
-        },
-        browserify: {
```
I'm not sure if there is anything still requiring this step.

radibit replied:
It should be needed especially in this Karma related configuration.

stefantds replied:
@radibit ok, thanks!

stefantds commented on source/index.ts:
```diff
@@ -59,7 +60,7 @@ export {
     importSymKeyFromBase64,
     exportSymKeyToBase64,
-    KEYTYPES as keyTypes,
+    D4LKeyTypes as D4LKeyTypes,
```
this is a small breaking change (which we cold easily revert if needed)

stefantds commented on source/key.ts:
```diff
@@ -0,0 +1,366 @@
+    PASSWORD_KEY = 'pk',
+    ATTACHMENT_KEY = 'ak',
+    TAG_ENCRYPTION_KEY = 'tek',
+    APP_PRIVATE_KEY = 'apriv',
```
here is another breaking change as I wanted to transform these values to an enum. it used to be:
```
    APP: {
        PRIVATE_KEY: 'apriv',
        PUBLIC_KEY: 'apub',
    },
    USER: {
        PRIVATE_KEY: 'upriv',
        PUBLIC_KEY: 'upub',
    },
```

ibihim replied:
Your idea adds structure :+1:

stefantds commented on source/key.ts:
```diff
@@ -0,0 +1,366 @@
+ * @param {AsymKeyTypes} type - type of the key pair (app/user)
+ * @returns {Promise<D4LAsymKeysPair>} Resolves to an object containing a public and a private key as d4lKey objects
+ */
+const generateAsymKeyPair = (type: AsymKeyTypes): Promise<D4LAsymKeysPair> =>
```
I changed the type of the parameter here as it was depending on the enum-objects that I removed (it was also wrongly documented):
```
@param {String} type - type of the key pair 'A' for App, 'U' for User
[...]
            exportKey(keyPair.publicKey, type.PUBLIC_KEY),
            exportKey(keyPair.privateKey, type.PRIVATE_KEY),
```

ibihim commented on source/key.ts:
```diff
@@ -0,0 +1,366 @@
+                {
+                    name: 'PBKDF2',
+                    salt,
+                    iterations: ITERATIONS,
```
Nit: Makes me wonder if it makes sense to have `ITERATIONS` is too generic as a name. Maybe defining a whole PBKDF2 object would be cool... or not? :smile:  Like in a JS file where you have a quick overview of all the ciphers we use... and maybe an indirection on top so we can easily... sorry... drifting into symbolic links :laughing:

stefantds replied:
yeah, makes sense. Will reconsider the whole const suggestions altogether in a future PR.

ibihim commented on source/key.ts:
```diff
@@ -0,0 +1,366 @@
+ */
+const exportPublicKeyToSPKI = (cryptoKey: CryptoKey): Promise<b64> =>
+    crypto.subtle.exportKey('spki', cryptoKey)
+        .then(SPKI => new Uint8Array(SPKI))
```
Nit: `SPKI` is not a constant so lower case?

ibihim reviewed (commented):
Nice effort!

radibit commented on package.json:
```diff
@@ -1,56 +1,59 @@
+        "@typescript-eslint/eslint-plugin": "^4.28.3",
+        "@typescript-eslint/parser": "^4.28.3",
+        "babel-eslint": "^8.2.6",
+        "babelify": "^9.0.0",
```
Do we still need `babelify` considering the changes at https://github.com/d4l-data4life/js-crypto/pull/18/files#diff-25bdb46a5428d318648ec9eb46a99c0268d8c96b672c46e2247a685bc3e384fbL20?

radibit commented on source/key.ts:
```diff
@@ -0,0 +1,366 @@
+            exportKey(
+                keyPair.publicKey, type === AsymKeyTypes.APP ?
+                D4LKeyTypes.APP_PUBLIC_KEY :
+                D4LKeyTypes.USER_PUBLIC_KEY) as Promise<D4LKeyPublic>,
```
This misalignment will be resolved as soon as we introduce Prettier to this project.

radibit reviewed (commented):
Great work on this PR especially considering the goal to keep it as lean as possible for a review and step-by-step code refactoring in series of PR 👏 

I've left several minor comments aside from what has been already addressed from the previous reviewer.

Would be nice to keep track of the breaking changes outside of the PRs so we can easily add them later to the final release. Maybe a wiki page or section in the readme?

As a follow up PR, I'd add Prettier to the project so we can unify its code formatting with the rest of the projects.

stefantds commented on source/index.ts:
```diff
@@ -39,12 +40,12 @@ import {
+const crypto = window.crypto || (window['msCrypto'] as Crypto);
 
-const hash = (message, alg = 'SHA-512') => crypto.subtle.digest(alg, Buffer.from(message))
+const hash = (message: ArrayBuffer, alg = 'SHA-512'): Promise<b64> => crypto.subtle.digest(alg, Buffer.from(message))
```
I agree, but it's not something that I introduced and not in scope for this refactoring. Will keep in mind for the next steps.

stefantds commented on source/utils.ts:
```diff
@@ -7,37 +8,39 @@ const textDecoder = new TextDecoder();
-            resolve(new Uint8Array(event.target.result));
+
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            resolve(new Uint8Array(event.target.result as any)); // TODO check the type mismatch here, remove "any"
```
I agree, but it's not something that I introduced and not in scope for this refactoring. Will keep in mind for the next steps.

stefantds commented on source/key.ts:
```diff
@@ -0,0 +1,366 @@
+    // 3. Usage of the keys. (http://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface-types)
+    crypto.subtle.generateKey(
+        RSA_OAEP,
+        true,
```
yeah, although on this we have more issues with how we handle the extractable property of the WebCrypto keys/ E.g. it is confusing and hidden that for some keys we use extractable while for others unextractable keys. This should be left to the client to decide, this library should have no idea how these keys are used.

stefantds replied:
(nevertheless, the suggestion improves a lot the readability so I will add it already)

stefantds commented:
@radibit thanks for the review, I addressed the comments (except the constants, as it goes beyond the scope of this PR in my opinion).
Please take a look again.
I still have to see why the build is failing, maybe you already know what is missing there?

radibit commented:
@stefantds Regarding the constants 👍 

About the failing build, it looks like the `typescript-eslint` might need to explicitly exclude of checking any files under `node_modules`. You can try adding a `.eslintignore` rule for that directory. Reference GH thread https://github.com/facebook/create-react-app/issues/8936#issuecomment-674060591.

stefantds commented:
@radibit the issue was actually what it was stated: typescript was not installed. I fixed the issue by adding typescript as a dev dependency. I missed that as I had typescript installed globally.

radibit reviewed (approved):
Looks good 🥇

On the TypeScript dependency, indeed it's a required one for any of the plugins that relay on it like in the case with linting a `*.ts` files 😅
